### ISSUE TITLE
[Feature Request] Ability to declare an enddate

### ISSUE THREAD

**Issue by Novalis15** [jagrosh/GiveawayBot#112]
Hey, i thought it would be cool to instead of just being able to declare an additive time period like `3H` for 3 Hours + the current datetime for the giveaway to end, to be able to just declare the point in time in which the giveaway should end.

E.g.: in my case i know that i want a giveaway which ends on the 10th October 2021 at 11 p.m., so it would be nice if i could just declare this date as the endpoint of the giveaway.

It would also help if we were able to accumulate these options: 3H 2M would then mean that the giveaway ends in 3 hours and 2 minutes.

**Comment by robinduerhager** [comment]
Hey, a colleague and i just had a chat about timezones and we were asking ourselves if it wouldn't be possible to input an enddate to the giveaway bot which calculates the offset based on that and the current datetime the user has (based on the users timezone)?

**Comment by jagrosh** [comment]
The bot doesn't know the user's timezone.

I believe that the Discord devs have mentioned adding a type of field for a date, which would likely be a requirement before this kind of feature was attempted.

**Comment by Dougley** [comment]
Why would the timezone matter in this context? The user wants the giveaway to end at X in their timezone presumably, so knowing exact timezones seems irrelevant.

**Comment by JetCrooked** [comment]
> Why would the timezone matter in this context? The user wants the giveaway to end at X in their timezone presumably, so knowing exact timezones seems irrelevant.

because if the bot receives (and by extension shows) a plain text end time then that time is wrong for anyone not in the giveaway host's timezone?

**Comment by Dougley** [comment]
> because if the bot receives (and by extension shows) a plain text end time then that time is wrong for anyone not in the giveaway host's timezone?

[It doesn't have to show a plain timestamp](https://discord.com/developers/docs/reference#message-formatting-timestamp-styles)

**Comment by JetCrooked** [comment]
reposting from <#585692140651151366>
![image](https://user-images.githubusercontent.com/30682557/186748104-0973ba36-59bc-470a-80a2-a8106ef45997.png)

**Comment by JetCrooked** [comment]
I am Confused

**Comment by Sanduhr32** [comment]
> Why would the timezone matter in this context? The user wants the giveaway to end at X in their timezone presumably, so knowing exact timezones seems irrelevant.

You contradict yourself. The users timezone matters, because it may not be equal to bot/host/UTC.
However you may be able to assume a hosting users timezone with their locale.

> If you absolutely need to implement timezones, you could always require UTC for exact datetimes. It would make this feature more of a power tool but I assume this was already intended as a power tool.

If timezones were implemented, they would internally always be converted to the bots host timezone (UTC in the best case)

> It doesn't have to show a plain timestamp

Afaik it shows a relative and an absolute UTC unix timestamp of the endtime. The client will format it according to the users local timezone.
(You can have timezone depending unix timestamps)



The implementation will be properly done regarding timezones, if it's done. I wouldn't expect anything else because everyone knows how delicate timezones are.

I can suggest javas DateTimeFormatter using a standard or custom pattern for parsing timestamps reliably with ease.
This includes timezone information. Javas modern time api also allows you to adjust given time in a timezone to a different timezone.

**Comment by jagrosh** [comment]
This is all related to creating a giveaway, so for example, the user wants to be able to input the end of the giveaway as: `August 30th 2022 5:00 PM`, or something like `Tomorrow at 4pm`. However, this exact time may be different for the user and for GiveawayBot, so GiveawayBot would _also_ need to know the user's timezone. Requiring the timezone in the prompt though drastically complicates things for the average user, who may not even know the technical name for their timezone.

The two ways I could see this being user-friendly are:
1. Discord adds a "date" type option to commands, which would bring up a calendar modal (like how the search calendar appears) which would provide a unix timestamp under the hood to the bot (which would not be ambiguous about timezone).
2. Utilize an existing system for allowing scheduling of giveaways. For example, the ability for a Discord event to have an attached giveaway that matches the event's time.

**Comment by jagrosh** [comment]
Just for further clarification, in line with the original post, the request is essentially the ability to do:
```
/gstart end:10th October 2021 at 11 p.m. winners:1 prize:Something
```

**Comment by Sanduhr32** [comment]
> The three ways I could see this being user-friendly are:
> 
>  1. Discord adds a "date" type option to commands, which would bring up a calendar modal (like how the search calendar appears) which would provide a unix timestamp under the hood to the bot (which would not be ambiguous about timezone).

I completely agree with this one, i don't know why they have not implemented this yet. This is requested since day 0 of slash command options i believe.

>  2. Utilize an existing system for allowing scheduling of giveaways. For example, the ability for a Discord event to have an attached giveaway that matches the event's time.

This is a dirty workaround for 1. imo and not worth going for.

>  3. Require a timezone in addition to the human-readable date (this option is more complicated and confusing for a lot of users).

Considering the average gb user this isn't an option either, please forgive the harsh categorization.

**Comment by Dougley** [comment]
> Discord adds a "date" type option to commands, which would bring up a calendar modal

I don't know why you're suggesting to reinvent the wheel by waiting for a modal type that might not even be in development. 

Like I suggested earlier, using RFC 1123 / RFC 822 and accepting `/gstart end:Thu, 25 Aug 2022 20:19:59 GMT winners:1 prize:Something` is a good solution to this. 

>

**Comment by bup0441** [comment]
> Requiring the timezone in the prompt though drastically complicates things for the average user, who may not even know the technical name for their timezone.

You could use PHP normalized timezones as well - 
``/gstart end:Thu, 25 Aug 2022 20:19:59 timezone:Europe/London winners:1 prize:token access``
It could even be supported by autocomplete for ease of access, and if you don't want to roll it into /gstart for fear of complication, turn it into a separate command called /gschedule or something

**Comment by jagrosh** [comment]
Personally I think requiring a very specific format is going to cause more confusion and support queries/complaints than the benefit. Also, it has been stated [here](https://github.com/discord/discord-api-docs/discussions/3581) that a date picker option type is in development.
